app/services/get_course_option_from_codes.rb

```diff
@@ -0,0 +1,65 @@
+class GetCourseOptionFromCodes
+  include ActiveModel::Validations
+  attr_accessor :provider_code, :course_code, :recruitment_cycle_year, :study_mode, :site_code
+  attr_accessor :provider, :course, :site, :course_option
+
+  validates_each :provider_code do |record, attr, value|
```

Comment:
validates_each is a bit confusing at first here! is there a way around using it? if not, maybe add something to the commit message?

Reply:
Yes, I'd imagine it's a little confusing at first, but weirdly this is most direct way I could find validating all attributes against each other and populating the service with specific validation errors. I've done something similar in `ChangeOfferForm`. I think the underlying issue is validation for each attribute depends on the values of the other attributes of the instance.

See also: https://books.google.co.uk/books?id=GL2kAwAAQBAJ&pg=PA244&lpg=PA244&dq=alternatives+to+validates_each+rails&source=bl&ots=fAcRulJioj&sig=ACfU3U1UyXLx_oPCLcQlh43P7KggLWNvsA&hl=en&sa=X&ved=2ahUKEwixztXo4KjoAhWCYcAKHa4MDIEQ6AEwBXoECAkQAQ#v=onepage&q=alternatives%20to%20validates_each%20rails&f=false

Reply:
The majority of this class is validation code which makes me think it might warrant a separate validator class.

I'm possibly missing the point but I don't quite understand the use of `validates_each` here, it's documented as a way to apply the same block of validation logic on multiple attributes: https://guides.rubyonrails.org/active_record_validations.html#validates-each whereas we are applying it to a single attribute in several distinct blocks.
I think if we need validation logic which is conditional based on several record attributes a custom validator would serve this purpose https://guides.rubyonrails.org/active_record_validations.html#custom-validators as `ActiveModel::Validator#validate` receives the record as an argument.

Reply:
Having discussed this, I think this isn't a significant issue, it's very particular validation for a very specific case and fine as is. :+1: